Issue 4027: Add Metric for Container Recovery Time

RaulGracia opened the pull request:
**Change log description**  
Added a new metric to keep track of container recovery times.

**Purpose of the change**  
Fixes #4027.

**What the code does**  
This PR adds a new metric (`segmentstore.container.recovery_time`) to track the time it takes for a Segment Store to perform a container recovery. This will be useful to create alerts if a recovery takes too long to complete.

**How to verify it**  
The new metric should be reported. All other tests must pass.

kevinhan88 commented:
Please add unit tests for this new metric.
How about we add some tags into the metric? e.g. container epoc? will it be beneficial for the purpose?

RaulGracia commented:
@kevinhan88 thanks for the comment. Actually, we want to how long are recoveries taking to spot any problem either in the Segment Stores or in the storage. The only tag that comes to my mind that could be interesting to add would the "host", to know where the Segment Store instance that did the recovery is placed.l But I'm not sure that such info is available at the point where we are reporting this metric.

kevinhan88 commented:
> @kevinhan88 thanks for the comment. Actually, we want to how long are recoveries taking to spot any problem either in the Segment Stores or in the storage. The only tag that comes to my mind that could be interesting to add would the "host", to know where the Segment Store instance that did the recovery is placed.l But I'm not sure that such info is available at the point where we are reporting this metric.

Cool, the host is the default tag applied to all the metrics, so seems you're ready to go

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:
```diff
@@ -39,6 +43,17 @@
+    /**
+     * Global metric to publish the time elapsed for the recovery of Segment Store containers.
+     */
+    @VisibleForTesting
```
I believe the only reason you made this an `AtomicReference` is because of testing. There is another way you can do this:
1. Keep this as a `static final` field.
2. Create a `protected` getter for it; make sure you only use this getter in your code.
3. In your test, create a subclass where you override the getter with whatever you want to supply there.


You need to optimize for production code, not for tests. It's OK if your test has a subclass that exposes some fields - it's much better than exposing them unnecessarily or adding unneeded complexity for all cases.

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:
```diff
@@ -313,4 +328,17 @@ public void truncate() {
+     *
+     * @param duration Time taken for a Segment Store instance to perform the recovery of containers.
+     */
+    public static void recoveryCompleted(long duration) {
```
This should report the recovery per container. Some containers may have longer recovery times since they may have many more operations to recover. For example, in some of our past system tests, 2 containers had a 50,000 operation backlog, while the other 2 had less than 1000. Their recovery times were proportional to this. If we report this as a global metric, the actual durations may get smoothened out (especially if we have a lot of containers) so we won't be able to detect any spikes).

andreipaduroiu commented:
@RaulGracia Any updates on this PR?

RaulGracia commented:
@andreipaduroiu I have been doing other pressing issues, but it is in my mind. Will try to do some progress this week, thanks!

RaulGracia commented:
@andreipaduroiu I have retaken this PR. I have changed the type of metric, as I see a "gauge per container" a better way to expose the recovery times of containers (i.e., it is clearer than a histogram, given that recovery events are expected to be rare, conversely to event writes, for instance, for which a histogram makes sense). Apart from that, I have addressed your comments and tested the code of this PR in a real cluster. As you can see, the new metric is being correctly reported:

<img width="957" alt="recovery_metric" src="https://user-images.githubusercontent.com/717112/71678930-9b254500-2d86-11ea-8305-9c089d1a1c4e.png">

I wrote data to a Stream with 12 segments, and as can be seen, 12 containers present much higher container recovery times after deleting a Segment store instance (all the containers are recovered).
